chain/subscriber.go

```diff
@@ -0,0 +1,517 @@
+		defer func() { cs.knownContracts[fcid] = cu.State }()
+
+		// update state from 'active' -> 'pending'
+		if rev == nil {
+			cu.State = api.ContractStatePending
+		}
+
+		// reverted renewal: 'complete' -> 'active'
+		if rev != nil {
+			cu.RevisionHeight = &cru.State.Index.Height
+			cu.RevisionNumber = &prevRev.revisionNumber
+			cu.Size = &prevRev.fileSize
+			if cu.State == api.ContractStateComplete {
+				cu.State = api.ContractStateActive
+				cs.logger.Infow("contract state changed: complete -> active",
+					"fcid", fcid,
+					"reason", "final revision reverted")
+			}
+		}
+
+		// reverted storage proof: 'complete/failed' -> 'active'
+		if resolved {
+			cu.State = api.ContractStateActive // revert from 'complete' to 'active'
+			if valid {
+				cs.logger.Infow("contract state changed: complete -> active",
+					"fcid", fcid,
+					"reason", "storage proof reverted")
+			} else {
+				cs.logger.Infow("contract state changed: failed -> active",
+					"fcid", fcid,
+					"reason", "storage proof reverted")
+			}
+		}
+	}
+
+	// v1 contracts
+	cru.ForEachFileContractElement(func(fce types.FileContractElement, rev *types.FileContractElement, resolved, valid bool) {
+		var r *revision
+		if rev != nil {
+			r = &revision{
+				revisionNumber: rev.FileContract.RevisionNumber,
+				fileSize:       rev.FileContract.Filesize,
+			}
+		}
+		prevRev := revision{
+			revisionNumber: fce.FileContract.RevisionNumber,
+			fileSize:       fce.FileContract.Filesize,
+		}
+		processContract(types.FileContractID(fce.ID), prevRev, r, resolved, valid)
+	})
+
+	// v2 contracts
+	cru.ForEachV2FileContractElement(func(fce types.V2FileContractElement, rev *types.V2FileContractElement, res types.V2FileContractResolutionType) {
+		var r *revision
+		if rev != nil {
+			r = &revision{
+				revisionNumber: rev.V2FileContract.RevisionNumber,
+				fileSize:       rev.V2FileContract.Filesize,
+			}
+		}
+		resolved := res != nil
+		valid := false
+		if res != nil {
+			switch res.(type) {
+			case *types.V2FileContractFinalization:
+				valid = true
+			case *types.V2FileContractRenewal:
+				valid = true
+			case *types.V2StorageProof:
+				valid = true
+			case *types.V2FileContractExpiration:
+				valid = fce.V2FileContract.Filesize == 0
+			}
+		}
+		prevRev := revision{
+			revisionNumber: fce.V2FileContract.RevisionNumber,
+			fileSize:       fce.V2FileContract.Filesize,
+		}
+		processContract(types.FileContractID(fce.ID), prevRev, r, resolved, valid)
+	})
+}
+
+func (cs *Subscriber) processChainApplyUpdateWallet(cau chain.ApplyUpdate) error {
+	return wallet.ApplyChainUpdates(cs.nextUpdate, cs.walletAddress, []chain.ApplyUpdate{cau})
+}
+
+func (cs *Subscriber) processChainRevertUpdateWallet(cru chain.RevertUpdate) error {
+	return wallet.RevertChainUpdate(cs.nextUpdate, cs.walletAddress, cru)
+}
+
+func (cs *Subscriber) sync(index types.ChainIndex) error {
+	cs.mu.Lock()
+	if cs.syncing {
+		cs.mu.Unlock()
+		return nil
```

Comment:
This is probably not safe anymore since it drops `sync` calls without actually syncing.